### PR TITLE
TETRIS-ADAPT-VQE

### PR DESCRIPTION
**Pull request by AlexandreF-1qbit**
Highlights:
- Implementation of TETRIS-ADAPT-VQE. Operators acting on different set of qubits can be added during the same ADAPT cycle.
- Filtered the "No variational gate found in the circuit." warning. During the very first cycle of ADAPT building, there is obviously no variational parameters as the ansatz is empty.
- Fix for #224. The optimizer is not called again when ADAPT has converged when `ADAPTSolver.iteration < ADAPTSolver.max_cycles`). I moved the rebuilding of the circuit at the end of the simulate block (no logic is needed).

### CODE REVIEW

File: tangelo/algorithms/variational/tests/test_tetris_adapt_vqe_solver.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_build_tetris_adapt(self):
+        """Try instantiating TETRISADAPTSolver with basic input."""
+
+        opt_dict = {"molecule": mol_H2_sto3g, "max_cycles": 15}
+        adapt_solver = TETRISADAPTSolver(opt_dict)
+        adapt_solver.build()
+
+    def test_single_cycle_adapt(self):
+        """Try instantiating TETRISADAPTSolver with basic input."""
```
Comment (ValentinS4t1qbit) [inline tangelo/algorithms/variational/tests/test_tetris_adapt_vqe_solver.py:31]: I think you forgot to modify your copy-paste here

File: tangelo/algorithms/variational/adapt_vqe_solver.py [hunk 2]
```diff
@@ -285,12 +290,28 @@ def rank_pool(self, pool_commutators, circuit, backend, tolerance=1e-3):
...
+        Args:
+            list of float: Operator gradients corresponding to self.pool_operators.
+            tolerance (float): Minimum value for gradient to be considered.
+
+        Returns:
+            list of int: Index (list of length=1) of the operator with the
+                highest gradient. If it is not bigger than tolerance, returns
+                an empty list.
+        """
+        sorted_op_indices = sorted(range(len(gradients)), key=lambda k: gradients[k])
```
Comment (ValentinS4t1qbit) [inline tangelo/algorithms/variational/adapt_vqe_solver.py:308]: The previous implementation looked for the maximum value in the gradient and returned its index.
This new implementation sorts / ranks all entries in the gradient and retrieves the last, corresponding to the largest value in there, which is more complicated.

What motivated this choice ? Was it related to stability ?

File: tangelo/algorithms/variational/adapt_vqe_solver.py [hunk 3]
```diff
@@ -285,12 +290,28 @@ def rank_pool(self, pool_commutators, circuit, backend, tolerance=1e-3):
...
-        max_partial = max(gradient)
+
+        return gradient
+
+    def choose_operator(self, gradients, tolerance=1e-3):
+        """Choose next operator to add according to the ADAPT-VQE algorithm.
+
+        Args:
+            list of float: Operator gradients corresponding to self.pool_operators.
+            tolerance (float): Minimum value for gradient to be considered.
```
Comment (ValentinS4t1qbit) [inline tangelo/algorithms/variational/adapt_vqe_solver.py:301]: `gradients` has not been documented

File: tangelo/algorithms/variational/tetris_adapt_vqe_solver.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    """
+
+    def choose_operator(self, gradients, tolerance=1e-3):
+        """Choose the next operator(s) to add according to the TETRIS-ADAPT-VQE
+        algorithm.
+
+        Args:
+            list of float: Operator gradients (absolute values) corresponding to
+                self.pool_operators.
+            tolerance (float): Minimum value for gradient to be considered.
```
Comment (ValentinS4t1qbit) [inline tangelo/algorithms/variational/tetris_adapt_vqe_solver.py:44]: `gradients` has not been documented

File: tangelo/toolboxes/operators/operators.py [hunk 5]
```diff
@@ -124,6 +124,24 @@ def get_max_number_hamiltonian_terms(self, n_qubits):
...
+        """Return a set of integers corresponding to qubit indices the qubit
+        operator acts on.
+
+        Returns:
+            set: Set of qubit indices.
+        """
+
+        qubit_indices = set()
+        for term in self.terms:
+            if not term:
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/operators/operators.py:138]: Why not just:

```
if term:
   indices = list(zip(*term))[0]
   qubit_indices.update(indices)
```

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
Good use of inheritance to define a new solver with minimum code and ensure it remains sync'd with changes in ADAPT.

Very little to say on this PR.

File: tangelo/algorithms/variational/adapt_vqe_solver.py (hunk 2 above)
Comment (AlexandreF-1qbit) [reply tangelo/algorithms/variational/adapt_vqe_solver.py:308]: Yeah, mainly stability, and I think this implementation is easier to debug in general. I can revert back to the original if you desire.

Comment (JamesB-1qbit) [reply tangelo/algorithms/variational/adapt_vqe_solver.py:309]: if you are willing to import numpy then this line can be replaced by.
```python3
sorted_op_indices = np.argsort(gradients)
```

Comment (ValentinS4t1qbit) [reply tangelo/algorithms/variational/adapt_vqe_solver.py:309]: @AlexandreF-1qbit Then I think it's ok to keep your implementation. I let you decide whether or not to implement James suggestion.

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you !

File: tangelo/algorithms/variational/adapt_vqe_solver.py (hunk 2 above)
Comment (AlexandreF-1qbit) [reply tangelo/algorithms/variational/adapt_vqe_solver.py:309]: I think I prefer not importing numpy for this one-time purpose in this file.

**Pull request merged**